tests/devices/test_default_qubit.py

```diff
@@ -1989,3 +1661,136 @@ def test_s_inverse():
 
         expected = np.array([1., -1.j]) / np.sqrt(2)
         assert np.allclose(dev.state, expected, atol=tol, rtol=0)
+
+@pytest.mark.parametrize("inverse", [True, False])
+class TestApplyOperationUnit:
+    """Unit tests for the internal _apply_operation method."""
+
+    def test_internal_apply_ops_case(self, inverse):
+        """Tests that if we provide an operation that has an internal
+        implementation, then we use that specific implementation."""
+        dev = qml.device('default.qubit', wires=1)
+
+        # Create a dummy operation
+        expected_test_output = np.ones(1)
+        paulix = lambda *args, **kwargs: expected_test_output
+
+        # Set the internal ops implementations dict
+        dev._apply_ops = {"PauliX": paulix}
+
+        test_state = np.array([1,0])
+        op = qml.PauliX(0) if not inverse else qml.PauliX(0).inv()
+
+        res = dev._apply_operation(test_state, op)
+        assert np.allclose(res, expected_test_output)
+
+    def test_diagonal_operation_case(self, inverse, mocker, monkeypatch):
+        """Tests the case when the operation to be applied is a
+        DiagonalOperation and the _apply_diagonal_unitary method is used."""
+        dev = qml.device('default.qubit', wires=1)
+        par = 0.3
+
+        test_state = np.array([1,0])
+        wires = 0
+        op = qml.PhaseShift(par, wires=wires) if not inverse else qml.PhaseShift(par, wires=wires).inv()
+        assert op.name not in dev._apply_ops
+
+        # Set the internal _apply_diagonal_unitary
+        history = []
+        mock_apply_diag = lambda state, matrix, wires: history.append((state, matrix, wires))
+        with monkeypatch.context() as m:
+            m.setattr(dev, "_apply_diagonal_unitary", mock_apply_diag)
+            assert dev._apply_diagonal_unitary == mock_apply_diag
+
+            dev._apply_operation(test_state, op)
+
+            res_state, res_mat, res_wires = history[0]
+
+            assert np.allclose(res_state, test_state)
+            assert np.allclose(res_mat, np.diag(op.matrix))
+            assert np.allclose(res_wires, wires)
+
+    def test_apply_einsum_case(self, inverse, mocker, monkeypatch):
+        """Tests applying a diagonal operation."""
+        dev = qml.device('default.qubit', wires=1)
+
+        test_state = np.array([1,0])
+        wires = 0
+
+        # Recreate the S gate, so that it's an example for a one-qubit gate
+        # that doesn't inherit from DiagonalOperation
```

Comment:
```suggestion
        # Recreate the S gate, so that it is an example for a one-qubit gate
        # that does not inherit from DiagonalOperation
```